### PR TITLE
Implement Bind and Execute processes of Postgres Extended Query protocol

### LINKED ISSUE #177: Implement Extended Query PostgreSQL Protocol
If I were asked to describe what is `"Extended Query PostgreSQL Protocol"`, I'd say it is `Prepared Statement` over wire.
As far as I understand (right now) query is sent in form of the string like `select * from table where col1 = $1;` or `insert into table values ($1, $2, $3);` and then data bindings negotiated between client and server.

Doc https://www.postgresql.org/docs/12/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY

Quick search over Materialize source code gave me https://github.com/MaterializeInc/materialize/blob/main/src/pgwire/src/protocol.rs#L114 as possible entry point to start getting info of how "prepared statement" state could be managed

### PR DESCRIPTION
**Pull request by silathdiir**
Closes #177 

#### Description

The [Postgres Extended Query](https://www.postgresql.org/docs/12/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY) includes three steps - `Parse`, `Bind` and `Execute`.
This PR implements the remain `Bind` and `Execute`.

#### Client output

For testing Extended Query, download `epgsql` and run the commands as below.

```
git clone git@github.com:epgsql/epgsql.git
cd epgsql
make

./rebar3 shell

1> {ok, C} = epgsql:connect("localhost", "user", "password", #{codecs => []}).
2> epgsql:squery(C, "create schema schema_name").
3> epgsql:squery(C, "create table schema_name.table_name (column_1 smallint, column_2 smallint)").

4> {ok, S1} = epgsql:parse(C, "prepared_stmt_name_1", "insert into schema_name.table_name values ($1, $2)", [int4, int4]).
5> {ok, S2} = epgsql:parse(C, "prepared_stmt_name_2", "update schema_name.table_name set column_1 = $1, column_2 = $2", [int4, int4]).

6> epgsql:bind(C, S1, "protal_name_1", [1, 1]).
7> epgsql:bind(C, S2, "protal_name_2", [2, 2]).

8> epgsql:execute(C, S1, "protal_name_1", 0).
9> epgsql:squery(C, "select * from schema_name.table_name").

10> epgsql:execute(C, S2, "protal_name_2", 0).
11> epgsql:squery(C, "select * from schema_name.table_name").
```

It seems that `sqlparser-rs` team did not want to add `$` as an Identifier to `PostgreSqlDialect`. 

https://github.com/ballista-compute/sqlparser-rs/issues/48
https://github.com/ballista-compute/sqlparser-rs/pull/243

We may need to keep `PreparedStatementDialect`.

### TODO

There are other remain issues needed to be implemented (in later PRs):

1. Only Parameter binding for `Select` and `Update` have been handled for now. `Select` and `Delete` need to be implemented.

2. Parameter binding is only implemented on simple `Expr`s. As `Expr::BinaryOp` and `Expr::UnaryOp` need to be implemented.

3. Returning `max_rows` of Protocol `Execute` need to be handled (after binding `Select`).

4. The Extended Query is only one part of feature Prepared Statement. We could also use Prepared Statement on Simple Query via SQL `Prepare`, `Execute` and `Deallocate`. As mentioned in https://github.com/ballista-compute/sqlparser-rs/pull/243.

### CODE REVIEW

**Comment by silathdiir** [comment]
@alex-dukhno Please help review if you have time.

**Comment by alex-dukhno** [comment]
> 4. The Extended Query is only one part of feature Prepared Statement. We could also use Prepared Statement via Simple Query via SQL Prepare, Execute and Deallocate. As mentioned in ballista-compute/sqlparser-rs#243.

Can you show an example how it could be used from user perspective?

> There are other remain issues needed to be implemented (in later PRs):
>
> 1. Only Parameter binding for Select and Update have been handled for now. Select and Delete need to be implemented.
>
> 2. Parameter binding is only implemented on simple Exprs. As Expr::BinaryOp and Expr::UnaryOp need to be implemented.
>
> 3. Returning max_rows of Protocol Execute need to be handled (after binding Select).

Can you please create issue for that? You should have access to create `enhancement`s. Thanks.

**Comment by silathdiir** [comment]
Hi @alex-dukhno, I got the information from this comment of materialize: https://github.com/MaterializeInc/materialize/blob/main/src/coord/src/session/statement.rs#L12.

For testing Prepared Statement via Simple Query on my local Mac, I started an instance via `brew services start postgresql`.
Connects it via `psql`, and inputs the below commands.

```
=# create schema schema_name;
=# create table schema_name.table_name (s1 int, s2 int);
=# \d schema_name.table_name;

=# prepare prepared_stmt_name (int, int) as insert into schema_name.table_name values($1, $2);
=# execute prepared_stmt_name(1, 1);
=# select * from schema_name.table_name;
```

For further testing with `git@github.com:epgsql/epgsql.git`, it seems that Postgres could recognize and use Prepared Statement from both sources.

I create a Prepared Statement via `Parse` of Extended Query. And execute it by Simple Query as below (`epgsql:squery("execute prepared_statement_name(1, 1)"`).

The only limitation is must invoking on the same session (Client socket `C` as below).

```
git clone git@github.com:epgsql/epgsql.git
cd epgsql
make

./rebar3 shell

1> {ok, C} = epgsql:connect("localhost", "user", "password", #{codecs => []}).
2> epgsql:squery(C, "create schema schema_name").
3> epgsql:squery(C, "create table schema_name.table_name (column_1 smallint, column_2 smallint)").

4> {ok, S} = epgsql:parse(C, "prepared_stmt_name", "insert into schema_name.table_name values ($1, $2)", [int4, int4]).

5> epgsql:squery(C, "execute prepared_stmt_name(1, 1)").
5> epgsql:squery(C, "select * from schema_name.table_name").
```

**Comment by silathdiir** [comment]
> Can you please create issue for that? You should have access to create `enhancement`s. Thanks.

@alex-dukhno Sure, I will list and create issues for `enhancement`s. Thanks.

File: src/protocol/src/sql_values.rs [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+        Ok(v) => v,
+        Err(_) => return Err(format!("Failed to parse SmallInt from: {}", s)),
+    };
+
+    Ok(PostgreSqlValue::Int64(v))
+}
+
+fn parse_bool_from_binary(buf: &[u8]) -> Result {
+    let len = buf.len();
+    if len != 1 {
```
Comment (alex-dukhno) [inline src/protocol/src/sql_values.rs:98]: Sorry, I can't see why condition is strict `!= 1`? Can you clarify a bit?

File: src/sql_engine/src/dml/select.rs [hunk 2]
```diff
@@ -45,107 +45,115 @@ impl<'sc> SelectCommand<'sc> {
...
-        } else {
-            self.session
-                .send(Err(QueryError::feature_not_supported(self.raw_sql_query.to_owned())))
-                .expect("To Send Query Result to Client");
-            Err(SystemError::runtime_check_failure("Feature Not Supported".to_owned()))
         }
     }
 
-    pub(crate) fn execute(&mut self) -> SystemResult<()> {
+    fn parse_select_input(&self) -> SystemResult<SelectInput> {
```
Comment (alex-dukhno) [inline src/sql_engine/src/dml/select.rs:156]: @AndrewBregger am I right that we'll have to use/move this function to `QueryProcssor`?
This is cool that we discover one more case to think for query planning 👍

File: src/node/src/node.rs [hunk 3]
```diff
@@ -75,16 +75,40 @@ pub fn start() {
...
                                 return;
                             }
+                            Ok(Ok(Command::Bind(
+                                portal_name,
+                                statement_name,
+                                param_formats,
+                                raw_params,
+                                result_formats,
+                            ))) => {
+                                match query_executor.bind_prepared_statement_to_portal(
```
Comment (alex-dukhno) [inline src/node/src/node.rs:85]: I think we need to introduce one more optional stage in query processing pipeline for prepared statement and move this function to separate struct.

No need to change anything for this comment.

File: src/sql_engine/src/lib.rs [hunk 4]
```diff
@@ -244,20 +302,92 @@ impl QueryExecutor {
             }
         };
     }
+
+    fn process_statement(&self, raw_sql_query: &str, statement: Statement) -> SystemResult<()> {
```
Comment (alex-dukhno) [inline src/sql_engine/src/lib.rs:306]: I suppose it is one more sign to reorganise `QueryExecutor` struct.

No need to change anything for this comment.

File: src/sql_engine/src/query/bind.rs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+use sqlparser::ast::{Assignment, Expr, Ident, Query, SetExpr, Statement, Value};
+use std::sync::Arc;
+
+type Result = std::result::Result<(), ()>;
+
+pub(crate) struct ParamBinder {
+    sender: Arc<dyn Sender>,
+}
+
+impl ParamBinder {
```
Comment (alex-dukhno) [inline src/sql_engine/src/query/bind.rs:26]: I think we need one more enhancement task to make this struct public and use in `node` module directly, at least this approach should be investigated.

No need to change anything for this comment.

File: src/protocol/src/sql_values.rs (hunk 1 above)
Comment (silathdiir) [reply src/protocol/src/sql_values.rs:98]: I think passing type `bool` and `char` in binary should be validated the length of `[u8]` to one. I referenced the code of `rust-postgres` (which is used by `materialize`): https://github.com/sfackler/rust-postgres/blob/master/postgres-protocol/src/types/mod.rs#L34

Comment (alex-dukhno) [reply src/protocol/src/sql_values.rs:98]: which `char` is in context? rust single item of a string or `char` sql type that have length?

Comment (alex-dukhno) [reply src/protocol/src/sql_values.rs:98]: I suppose `char` and `varcher` should be parsed in the same way and `len != 1` is ok for `bool`

Comment (silathdiir) [reply src/protocol/src/sql_values.rs:98]: Is the type `Char` is handled as a single character?

https://github.com/sfackler/rust-postgres/blob/master/postgres-protocol/src/types/mod.rs#L75

Comment (silathdiir) [reply src/protocol/src/sql_values.rs:98]: Sorry. I mean just the SQL type `Char`. It is handled as a single byte when passing in binary (not as `Varchar`).

**Comment by silathdiir** [comment]
Hi @alex-dukhno , I fixes the issues you mentioned in the code review. I list the fixes to avoid missing.

1. Fixes `enum Command`s to have name fields (as `FrontendMessage`).

2. Moves `decode` and `parse` functions from `sql_values.rs` to `sql_types.rs` as funtions of PostgreSqlType.

3. Removes the one byte limitation of `Char(n)` SQL type.

And I also merge the file `src/sql_engine/src/dml/select.rs` manually, please help check and review the fixes, and point out if I missed some ones. Thanks.

**Review by alex-dukhno — APPROVED** [review]

**Pull request merged**